features/nfc/doc/nfc_design.md

```diff
@@ -0,0 +1,424 @@
+* A Type Name Format indicator - indicates which namespace the type field belongs to ('Well-known NDEF types', MIME, Absolute URI, etc.)
+* A type field
+* An optional ID field
+* The record's value
+
+All arrays are passed by reference (no copy made).
+
+```cpp
+static bool parse(const uint8_t* buffer, size_t max_sz)
+ssize_t build(const uint8_t* buffer, size_t max_sz)
+
+uint8_t tnf()
+void set_tnf(uint8_t tnf)
+
+const uint8_t* type() const
+size_t type_size() const
+void set_type(const uint8_t* type, size_t type_size)
+
+const uint8_t* id() const
+size_t id_size() const
+void set_id(const uint8_t* id, size_t id_size)
+
+const uint8_t* value() const
+size_t value_size() const
+void set_value(const uint8_t* type, size_t type_size)
+```
+
+**Helpers**
+
+We will provide multiple helpers to make it easy to create/parse common record types:
+* URI
+* Text
+* Smart Poster
+* MIME data
+
+For instance, the `URIRecord`'s class API is as follows:
+```cpp
+static bool is_uri_record(const NDEFRecord& record)
+static URIRecord as_uri_record(const NDEFRecord& record)
+
+uri_prefix_t uri_prefix() const
+void set_uri_prefix(uri_prefix_t prefix)
+
+bool get_uri(char* uri, size_t max_sz) const
+size_t uri_size() const
+void set_uri(const char* uri)
+
+bool get_full_uri(char* uri, size_t max_sz) const
+size_t full_uri_size() const
+void set_full_uri(const char* uri)
+```
+
+This includes some helper classes to check whether a record is an URI record, and if so to construct an `URIRecord` instance from a `NDEFRecord`.
+
+In this case buffers are copied to account for the NULL-terminator character that is not present in the underlying byte buffer.
+
+## HAL APIs
+
+### NFC EEPROM API
+
+The one HAL API that will have to be implemented by vendors to make use of the `NFCEEPROM` class are the following virtual methods.
+
+From the upper layer's point of view, the EEPROM is a byte array that can be read from/written to. Long operations (reads, writes, erasures) must happen asynchronously. Booleans indicate whether a particular operation was succesful. Encoding is handled by the upper layer.
+
+Address 0 means the start of the NDEF buffer (not necessarily at address 0 in the EEPROM). 
+
+When a buffer is passed to the backend, the reference remains valid till the corresponding event is called.
+
+The `backend_set_size()` command is called to change the size of the buffer (within the limits set by `backend_get_max_size()`). Inversely that buffer size can be read by `backend_get_size()`.
+
+```cpp
+void backend_reset()
+size_t backend_get_max_size() const
+void backend_read_bytes(uint32_t address, size_t count)
+void backend_write_bytes(uint32_t address, const uint8_t* bytes, size_t count)
+void backend_set_size(size_t count)
+void backend_get_size()
+void backend_erase_bytes(uint32_t address, size_t size) 
+```
+
+The following events must be called to signal completion of long operations:
+```cpp
+void on_backend_has_read_bytes(bool success, const uint8_t* bytes)
```

Comment:
I think we can do better naming wise. I'm not sure that the `on_backend` prefix is required there.
Also I think it can be valuable to get a more descriptive error code.

Note: Shouldn't we get the size of the bytes read in the callback ?